Add Contributing.md (#2978)
apple/swift-package-manager: Documentation/Contributing.md

> ## Code Contributions

Reviewer: ```suggestion
# Contributing to Swift Package Manager
```

There are several types of contributions one can make. Bug fixes, documentation and enhancements that do not materially change the user facing semantics of Swift Package Manager should be submitted directly as PR. 

Larger changes that do materially change the semantics of Swift Package Manager (e.g. changes to the manifest format or behavior) are required to go through [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md).
To see how previous evolution decisions for SwiftPM have been made check out https://github.com/apple/swift-package-manager/blob/main/Documentation/Internals/PackageManagerCommunityProposal.md.  

For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/)  

> ### Ways to contribute

Reviewer: ```suggestion
## Reporting issues
```

Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  
JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).

> ### Requirements

Reviewer: ```suggestion
## Development environment
```

You have multiple ways to setup your development environment, here we will focus on two:  
* [Using Xcode](#using-xcode)
* [Using a standalone Swift toolchain](#using-standalone)

<a id="using-xcode">*A) _Use Xcode to setup what you need_*.</a>  
Xcode is only available for macOS.

1. Install latest Xcode - [https://developer.apple.com/xcode](https://developer.apple.com/xcode/)
2. Confirm you have the latest SwiftPM:
```
$> swift package --version
Swift Package Manager - Swift 5.3.0
```
3. Confirm you have the latest Swift version:
```
$> swift --version
Apple Swift version 5.3
```

<a id="using-standalone">*B) _Use standalone Swift toolchain_*</a>:  

> Download the toolchain from https://swift.org/download/ (or use tools like swiftenv) and confirm its installed correctly like step 2 and 3.

Reviewer: ```suggestion
Download the toolchain from https://swift.org/download/ (or use tools like swiftenv) and confirm its installed correctly as described above.
```

> ### <a name="getting-started">Getting Started</a>

Reviewer: ```suggestion
## <a name="getting-started">Getting Started</a>
```

1. Pull the SwiftPM repository:
```
git clone https://github.com/apple/swift-package-manager.git
```
2. Run your first build:
```
$> cd swift-package-manager
$> swift build
```
Make sure the build did not fail.  

A `.build/` folder will be generated and it should have inside a similar structure (including build binaries):
```
artifacts/
checkouts/
debug/
repositories
x86_64-apple-macosx
debug.yaml
manifest.db
workspace-state.json 
```
Binaries (in the example above) are in `x86_64-apple-macosx/`.  
If you need to build the generated binaries, run `swift-build` in inside `.build/`:
```
./.build/x86_64-apple-macosx/debug/swift-build
```

### Code Editor

> If you are contributing using macOS, the best option is to use Xcode to build and run test SwiftPM. 

Reviewer: move this sentence in the "development environment" section and remove the "code editor" title

> ### Troubleshooting

Reviewer: lets move the troubleshooting section to the bottom of the document

* If during `swift build` you encounter this error:
```
/../apple-repos/swift-package-manager/.build/checkouts/swift-driver/Sources/SwiftDriver/Explicit Module Builds/InterModuleDependencyGraph.swift:102:3: error: unknown attribute '_spi'
  @_spi(Testing) public var isFramework: Bool
  ^
```
Make sure you are using SwiftPM 5.3
```
$> swift package --version
Swift Package Manager - Swift 5.3.0
```
* If during `swift build` you encounter this error:
```
/../swift-package-manager/Sources/PackageLoading/Target+PkgConfig.swift:84:36: error: type 'PkgConfigError' has no member 'prohibitedFlags'
            error = PkgConfigError.prohibitedFlags(filtered.unallowed.joined(separator: ", "))
                    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~
```
Make sure to update your TSC (Tools Support Core):
```
swift package update
```
Alternatively, if you are using Xcode, you can update to the latest version of all packages:  
**Xcode App** > *File* > *Swift Packages* > *Update to Latest Package Versions*

> ### Commits / PRs

Reviewer: ```suggestion
## Commits / PRs
```

1. Fork: https://github.com/apple/swift-package-manager
2. Clone a working copy of your fork
3. Create a new branch
4. Make your code changes
5. Commit (include the Radar link or JIRA issue id in the commit message if possible and a description your changes)
6. Push the commit / branch to your fork
7. Make a PR from your fork / branch to `apple: main`
8. Leave a new comment to trigger smoke tests: `@swift-ci please smoke test`
9. Reviewers are going to be automatically added to your PR
10. Merge pull request when you received approval from the reviewers (one or more)

> ### Community and Support

Reviewer: ```suggestion
## Community and Support
```

If you want to connect with the Swift community you can:
* Use Swift Forums: [https://forums.swift.org/c/development/SwiftPM](https://forums.swift.org/c/development/SwiftPM)
* Contact the CODEOWNERS: https://github.com/apple/swift-package-manager/blob/main/CODEOWNERS
(mailing lists are usually the best place to go for help: [code-owners@swift.org](mailto:code-owners@swift.org), [conduct@swift.org](mailto:conduct@swift.org), [swift-infrastructure@swift.org](mailto:swift-infrastructure@swift.org)

> ### Additional Links

Reviewer: ```suggestion
## Additional resources
```

# Swift Package Manager: Quick Start
Swift Package Manager (SwiftPM) is a tool for building, testing and managing Swift project dependencies.  
In order to use it you will need Swift 3.0 or greater.  
SwiftPM is also included in Xcode 8.0 and above.  
For usage getting started: [https://swift.org/getting-started/#using-the-package-manager](https://swift.org/getting-started/#using-the-package-manager)  

> For overview and examples: [https://swift.org/package-manager](https://swift.org/package-manager/)

Reviewer: This paragraph is redundant in this context, I would start this  document with the text under "Code Contributions" section

## Code Contributions

> Everyone is welcome to contribute to SwiftPM, submitting fixes, enhancement etc.  

Reviewer: We should add some text here about the different types of changes folks can make, something along the lines of

```markdown
There are several types of contributions one can make. Bug fixes, documentation and enhancements that do not materially change the user facing semantics of Swift Package Manager should be submitted directly as PR. 

Larger changes that do materially change the semantics of Swift Package Manager (e.g. changes to the manifest format or behavior) are required to go through [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md).
To see how previous evolution decisions for SwiftPM have been made check out https://github.com/apple/swift-package-manager/blob/main/Documentation/Internals/PackageManagerCommunityProposal.md

For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/) 
```

Find out how previous coding decisions for SwiftPM evolution have been made: https://github.com/apple/swift-package-manager/blob/main/Documentation/Internals/PackageManagerCommunityProposal.md

### Requirements

> You have multiple ways to setup your development environment, here we will focus on 2:  

Reviewer: ```suggestion
You have multiple ways to setup your development environment, here we will focus on two:  
```

> A) *[Using Xcode](#using-xcode)* or B) [Using *the standalone Swift toolchain*](#using-standalone).

Reviewer: ```suggestion
* [Using Xcode](#using-xcode)
* [Using a standalone Swift toolchain](#using-standalone).
```

<a id="using-xcode">*A) _Use Xcode to setup what you need_*.</a>
Xcode is only available for macOS.

> 1. Install Xcode 12 - [https://developer.apple.com/xcode](https://developer.apple.com/xcode/)

Reviewer: ```suggestion
1. Install latest Xcode - [https://developer.apple.com/xcode](https://developer.apple.com/xcode/)
```

> 2. Make sure you have at least SwiftPM 5.3:

Reviewer: ```suggestion
2. Confirm you have the latest SwiftPM:
```

```
$> swift package --version
Swift Package Manager - Swift 5.3.0
```

> 3. Make sure you have at least Swift 5.3:

Reviewer: ```suggestion
3. Confirm you have the latest Swift version:
```

```
$> swift --version
Apple Swift version 5.3
```

> If you were able to do and verify the steps above, go to [**Getting Started**](#getting-started)

Reviewer: this does not seem to be related?

> <a id="using-standalone">*B) _Use standalone Swift toolchain</a>: 2a) [On macOS](#on-macos) or 2b) [On Linux](#on-linux)_*.

Reviewer: I believe this should be simplified to say download the toolchain from https://swift.org/download/ (or use tools like swiftenv) and confirm its installed correctly like step 2 and 3 ^^

Procedure valid for macOS and Linux.

1. Pull the Swift repository:
```
git clone https://github.com/apple/swift.git
```
- 2 a. <a id="on-macos">On macOS</a>
```
PATH/TO/REPO/swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA
```
- 2b. <a id="on-linux">On Linux</a>
```
PATH/TO/REPO/swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
```

### <a name="getting-started">Getting Started</a>
1. Pull the SwiftPM repository:
```
git clone https://github.com/apple/swift-package-manager.git
```
2. Run your first build:
```
$> cd swift-package-manager
$> swift build
```
If the build process ends with exit code 0, the build is successful (we have an Enhancement Radar to implement a message for successful build and a short output on where the generated binaries are: rdar://69970428).  

> After a successful build (currently), you should see something like this:

Reviewer: I think ^^ can be removed / changes to say "make sure the build did not fail"

manifest.db
workspace-state.json 
```
Binaries (in the example above) are in `x86_64-apple-macosx/`.  
If you need to build the generated binaries, run `swift-build` in inside `.build/`:
```
./.build/x86_64-apple-macosx/debug/swift-build
```

### Code Editor
If you are contributing using macOS, the best option is to use Xcode to build and run test SwiftPM. 

### Troubleshooting
* If during `swift build` you encounter these outputs:
```
/../swift-package-manager/Sources/SPMTestSupport/misc.swift:93:35: warning: parameter 'file' with default argument '#file' passed to parameter 'file', whose default argument is '#filePath'
        XCTFail("\(error)", file: file, line: line)
                                  ^
/../swift-package-manager/Sources/SPMTestSupport/misc.swift:37:26: note: did you mean for parameter 'file' to default to '#filePath'?
    file: StaticString = #file,
                         ^~~~~
                         #filePath
/../swift-package-manager/Sources/SPMTestSupport/misc.swift:93:35: note: add parentheses to silence this warning
        XCTFail("\(error)", file: file, line: line)
                                  ^
                                  (   )

> ```

Reviewer: this can be removed ^^ as its a temporary issue

Do not worry, since those are known warnings that will be addressed at some point.  
Warnings differ depending on the platform and they can be seen from time to time due the amount of contributions.  
Our goal is to constantly monitor warnings and work on fix them (even if they are not affecting a successful implementation).
* If during `swift build` you encounter this error:
```
/../apple-repos/swift-package-manager/.build/checkouts/swift-driver/Sources/SwiftDriver/Explicit Module Builds/InterModuleDependencyGraph.swift:102:3: error: unknown attribute '_spi'
  @_spi(Testing) public var isFramework: Bool
  ^
```
Make sure you are using SwiftPM 5.3
```
$> swift package --version
Swift Package Manager - Swift 5.3.0
```
* If during `swift build` you encounter this error:
```
/../swift-package-manager/Sources/PackageLoading/Target+PkgConfig.swift:84:36: error: type 'PkgConfigError' has no member 'prohibitedFlags'
            error = PkgConfigError.prohibitedFlags(filtered.unallowed.joined(separator: ", "))
                    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~
```

> Make sure to update your TSC (Tools Support Core):

Reviewer: hmm TSC is a dependency so should be fetched properly, so not sure about this one

Reviewer: ~~I guess we could say run "swift package update"~~

```
swift package update
```
Alternatively, if you are using Xcode, you can update to the latest version of all packages:  
**Xcode App** > *File* > *Swift Packages* > *Update to Latest Package Versions*

### Find your way to contribute
Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  
JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).

> 

Reviewer: lets move this to the top of the doc along side the "ways to contribute" section suggested above

### Commits / PRs
1. Fork: https://github.com/apple/swift-package-manager
2. Clone a working copy of your fork
3. Create a new branch
4. Make your code changes
5. Commit (include the Radar link or JIRA issue id in the commit message if possible and a description your changes)
6. Push the commit / branch to your fork
7. Make a PR from your fork / branch to `apple: main`
8. Leave a new comment to trigger smoke tests: `@swift-ci please smoke test`
9. Reviewers are going to be automatically added to your PR

> 10. Merge pull request when you received approval from the reviewers (one or more)

Reviewer: <3

### Community and Support
If you want to connect with the Swift community you can:

> * Use Swift Forums: [https://forums.swift.org](https://forums.swift.org/)

Reviewer: should we point to the swiftpm specific area

* Contact the CODEOWNERS: https://github.com/apple/swift-package-manager/blob/main/CODEOWNERS
(mailing lists are usually the best place to go for help: [code-owners@swift.org](mailto:code-owners@swift.org), [conduct@swift.org](mailto:conduct@swift.org), [swift-infrastructure@swift.org](mailto:swift-infrastructure@swift.org)

### Additional Links
* Official Apple GitHub
[https://github.com/apple](https://github.com/apple)
* Swift Package Manager GitHub

> [https://github.com/apple/swift-package-manager](https://github.com/apple/swift-package-manager)

Reviewer: these two links  ^^ are redundant in this context

* Setup Development

> [https://github.com/apple/swift-package-manager/blob/main/Documentation/Development.md](https://github.com/apple/swift-package-manager/blob/main/Documentation/Development.md)

Reviewer: is there anything in Development.md that this doc does not cover? If not we should remove Development.md otherwise we should bring in the missing pieces and remove Development.md

Reviewer: @federicobucchi was ^^ adressed?

Author: I brought everything that was missed from Development.md and I removed Development.md

Larger changes that do materially change the semantics of Swift Package Manager (e.g. changes to the manifest format or behavior) are required to go through [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md).

To see how previous evolution decisions for SwiftPM have been made and have some direction for the development of future features please check out the [Community Proposals](https://forums.swift.org/tag/packagemanager).

For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/).

## Reporting issues
* [SwiftPM JIRA Bug Tracker](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22)
* [Guide for filing quality bug reports](https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug)

Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  

> JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).

Reviewer: L14 and L15 are now redundant as we cleaned them up in L11 and L12

> ## Development environment

Reviewer: here is an improved version we can use for this while section:

```
## Development environment

First, clone a copy of SwiftPM code from https://github.com/apple/swift-package-manager. 

If you are preparing to make a contribution you should fork the repository first and clone the fork which will make opening Pull Requests easier. See "Creating Pull Requests" section below.

SwiftPM is typically built with a pre-existing version of SwiftPM present on the system, but there are multiple ways to setup your development environment:

### Using Xcode (Easiest)

1. Install Xcode from [https://developer.apple.com/xcode](https://developer.apple.com/xcode) (including betas!).
2. Verify the expected version of Xcode was installed.
3. Open SwiftPM's `Package.swift` manifest with Xcode, and use Xcode to edit the code, build, and run the tests.

### Using the Command Line

If you are using macOS and have Xcode installed, you can use the command line even without downloading and installing a toolchain, otherwise, you first need to doownload and install one.

#### Installing the toolchain

1. Download a toolchain from https://swift.org/download/
2. Install it and verify the expected version of the toolchain was installed:

  **macOS**
  ```bash
  $> export TOOLCHAINS=swift
  ```

  Verify that we're able to find the swift compiler from the installed toolchain.
  ```bash
  $> xcrun --find swift
  /Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift
  $> swift package --version
  Swift Package Manager - Swift 5.3.0
  $> swift --version
  Apple Swift version 5.3
  ```

  **Linux**
  ```bash
  $> export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
  ```
  Verify that we're able to find the swift compiler from the installed toolchain.
  ```bash
  $> which swift
  /path/to/swift-toolchain/usr/bin/swift
  $> swift package --version
  Swift Package Manager - Swift 5.3.0
  $> swift --version
  Apple Swift version 5.3
  ```
  
Note:  Alternatively use tools like [swiftenv](https://github.com/kylef/swiftenv) that help manage toolchains versions.

#### Building

```bash
$> swift build
```

A successful build will create a `.build/` directory with the following approximate structure:
```bash
artifacts/
checkouts/
debug/
repositories
x86_64-apple-macosx
```

Binary artifacts are located in `x86_64-apple-macosx/` when building on macOS, 
or the equivalent on other architectures and operating systems.

These binaries can be used to test the code modification. For example, to test the `swift package init` and `swift build` commands from the new SwiftPM artifacts in `.build/`:

```bash
$> cd /tmp && mkdir hello && cd hello
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
```

#### Testing

```bash
$> swift test
```

to run a single test:

```bash
$> swift test --filter PackageGraphTests.DependencyResolverTests/testBasics
```

Or another example, to run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases:

```bash
$> swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
```

To run the performance tests, enable them with an ENV variable:

```bash
$> export TSC_ENABLE_PERF_TESTS=1
$> swift test -c release --filter PerformanceTests
```

### Using the bootstrap script

The bootstrap script is designed for building SwiftPM on systems that do not have Xcode or a toolchain installed.
It is used on bare systems to bootstrap the Swift toolchain (including SwiftPM), and as such not typically used outside the Swift team.

the bootstrap script requires having [CMake](https://cmake.org/) and [Ninja](https://ninja-build.org/) installed. 
Please refer to the [Swift project repo](https://github.com/apple/swift/blob/master/README.md#macos) for installation instructions.

1. Clone [llbuild](https://github.com/apple/swift-llbuild) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/apple/swift-llbuild llbuild
  ```

  Note: Make sure the directory for llbuild is called "llbuild" and not
      "swift-llbuild".

2. Clone [Yams](https://github.com/jpsim/yams) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/jpsim/yams
  ```

3. Clone [swift-driver](https://github.com/apple/swift-driver) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/apple/swift-driver
```

4. Clone [swift-argument-parser](https://github.com/apple/swift-argument-parser) beside the SwiftPM directory and check out tag with the [latest version](https://github.com/apple/swift-argument-parser/tags).

  For example, if the latest tag is 0.3.1:
  ```bash
  $> git clone https://github.com/apple/swift-argument-parser --branch 0.3.1
  ```

#### Building

```bash
$> Utilities/bootstrap build
```

See "Using the Command Line / Building" section above for more information on how to test the new artifacts.

#### Testing

```bash
$> Utilities/bootstrap test
```

## Testing locally

Before submitting code modification as Pull Requests, test locally across the supported platforms and build variants.

1. If using Xcode, run all the unit tests and verify they pass.

2. If using the Command Line, run all the unit tests and verify they pass.

  ```bash
  $> swift test
  ```

3. Optionally: Test with the bootstrap script as well.

  ```bash
  $> Utilities/bootstrap test
  ```
  
When developing on macOS and need to test on Linux, install [Docker](https://www.docker.com/products/docker-desktop) and use the following commands:

```bash
$> Utilities/Docker/docker-utils build # will build an image with the latest Swift snapshot
$> Utilities/Docker/docker-utils bootstrap # will bootstrap SwiftPM on the Linux container
$> Utilities/Docker/docker-utils run bash # to run an interactive Bash shell in the container
$> Utilities/Docker/docker-utils swift-build # to run swift-build in the container
$> Utilities/Docker/docker-utils swift-test # to run swift-test in the container
$> Utilities/Docker/docker-utils swift-run # to run swift-run in the container
```    

```

SwiftPM is typically built with a pre-existing version of SwiftPM present on the system, but there are have multiple ways to setup your development environment:

### Using Xcode (Easiest)

1. Install Xcode from [https://developer.apple.com/xcode](https://developer.apple.com/xcode) (including betas!).

> 2. Open SwiftPM's `Package.swift manifest with Xcode.

Reviewer: remove this line, we are just talking about setting up the environment here. we could replace it with an explanation on how to verify the right Xcode is installed

### Using a Swift toolchain

1. Download a toolchain from https://swift.org/download/

> 2. Confirm the toolchain is installed correctly:

Reviewer: Confirm -> Verify

*macOS*
```bash
$> export TOOLCHAINS=swift
```

Verify that we're able to find the swift compiler from the installed toolchain.
```bash
$> xcrun --find swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

*Linux*
```bash
$> export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
```
Verify that we're able to find the swift compiler from the installed toolchain.
```bash
$> which swift
/path/to/swift-toolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

> 3. Alternatively use tools like swiftenv that help manage toolchains versions.

Reviewer: remove the number 3, this is a note:

```
Note:  Alternatively use tools like [swiftenv](https://github.com/kylef/swiftenv) that help manage toolchains versions.
```

8. Push the commit / branch to your fork
9. Make a PR from your fork / branch to `apple: main`
10. While creating your PR, make sure to follow the PR Template providing information about the motivation and highlighting the changes.
11. Reviewers are going to be automatically added to your PR
12. Merge pull request when you received approval from the reviewers (one or more)

## Using Continuous Integration
SwiftPM uses [swift-ci](https://ci.swift.org) infrastructure for its continuous integration testing. The bots can be triggered on pull-requests if you have commit access. Otherwise, ask one of the code owners to trigger them for you.

Run tests with the trunk compiler and other projects. This is **required** before
a pull-request can be merged.

```
@swift-ci please smoke test
```

Run just the self-hosted tests. This has fast turnaround times so it can be used
to get quick feedback.

Note: Smoke tests are still required for merging pull-requests.

> ## Using Custom Swift Compilers

Reviewer: lets add an "advanced" title and put this under it

SwiftPM needs the Swift compiler to parse `Package.swift` manifest files and to
compile Swift source files. You can use the `SWIFT_EXEC` and `SWIFT_EXEC_MANIFEST`
environment variables to control which compiler to use for these operations.

`SWIFT_EXEC_MANIFEST`: This variable controls which compiler to use for parsing
`Package.swift` manifest files. The lookup order for the manifest compiler is:
`SWIFT_EXEC_MANIFEST`, `swiftc` adjacent to the `swiftpm` binaries, then `SWIFT_EXEC`

`SWIFT_EXEC`: This variable controls which compiler to use for compiling Swift
sources. The lookup order for the sources' compiler is: `SWIFT_EXEC`, then `swiftc` adjacent
to `swiftpm` binaries. This is also useful for Swift compiler developers when they
want to use a debug compiler with SwiftPM.

```bash
$> SWIFT_EXEC=/path/to/my/built/swiftc swift build
```

> ## Overriding the Path to the Runtime Libraries

Reviewer: put under advanced as well

environment variables to control which compiler to use for these operations.

`SWIFT_EXEC_MANIFEST`: This variable controls which compiler to use for parsing
`Package.swift` manifest files. The lookup order for the manifest compiler is:
`SWIFT_EXEC_MANIFEST`, `swiftc` adjacent to the `swiftpm` binaries, then `SWIFT_EXEC`

`SWIFT_EXEC`: This variable controls which compiler to use for compiling Swift
sources. The lookup order for the sources' compiler is: `SWIFT_EXEC`, then `swiftc` adjacent
to `swiftpm` binaries. This is also useful for Swift compiler developers when they
want to use a debug compiler with SwiftPM.

```bash
$> SWIFT_EXEC=/path/to/my/built/swiftc swift build
```

### Overriding the Path to the Runtime Libraries
SwiftPM computes the path of its runtime libraries relative to where it is
installed. This path can be overridden by setting the environment variable
`SWIFTPM_PD_LIBS` to a directory containing the libraries, or a colon-separated list of
absolute search paths. SwiftPM will choose the first
path which exists on disk. If none of the paths are present on disk, it will fall
back to built-in computation.

> ## Making changes in TSC targets

Reviewer: ```suggestion
### Making changes in TSC targets
```

```bash
$> xcrun --find swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

*Linux*
```bash
$> export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
```
Verify that we're able to find the swift compiler from the installed toolchain.
```bash
$> which swift
/path/to/swift-toolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

3. Alternatively use tools like swiftenv that help manage toolchains versions.

### Using the Swift Compiler Build Script

Follow these instructions to get the Swift sources and then execute the build-script using swiftpm preset:

*macOS*
```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA
```

*Linux*
```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
```

3. Clone [llbuild](https://github.com/apple/swift-llbuild) beside the package manager directory.

```bash
$> git clone https://github.com/apple/swift-llbuild llbuild

> $> ls

Reviewer: ls line seems redundant

swiftpm llbuild
```

Note: Make sure the directory for llbuild is called "llbuild" and not
    "swift-llbuild".

4. Clone [Yams](https://github.com/jpsim/yams) beside the package manager directory.

```bash
$> git clone https://github.com/jpsim/yams
```

5. Clone [swift-driver](https://github.com/apple/swift-driver) beside the package manager directory.

```bash
$> git clone https://github.com/apple/swift-driver
```

> 6. Clone [swift-argument-parser](https://github.com/apple/swift-argument-parser) beside the package manager directory and check out tag 0.3.0.

Reviewer: the version number should be the one specified in Package.swift so maybe instead of hard coding it here, say so and add link to the file

Author: Does this not referring to the tag version?

https://github.com/apple/swift-argument-parser/tags

```bash
$> git clone https://github.com/apple/swift-argument-parser --branch 0.3.0
```

7. Build the Swift Package Manager.

```bash
$> cd swiftpm
$> Utilities/bootstrap build
```

 Note: The bootstrap script requires having [CMake](https://cmake.org/) and [Ninja](https://ninja-build.org/) installed. Please refer to the [Swift project repo](https://github.com/apple/swift/blob/master/README.md#macos) for installation instructions.

This command builds the Package Manager inside the `.build/` directory.
    Run the bootstrap script to rebuild after making a change to the source
    code.

> <a id="self-hosting">*D) _Self Hosting_*</a>:  

Reviewer: This "D" section is redundant, remove it

```

Note: PackageDescription v4 is not available when developing using this method.

This method can also used be used for performance testing. Use the following
command run SwiftPM's performance tests:

```bash
$> export TSC_ENABLE_PERF_TESTS=1
$> swift test -c release --filter PerformanceTests
```

<a id="swift-compiler-build-script">*E) _Using the Swift Compiler Build Script_*</a>:  

Follow [these](https://github.com/apple/swift#getting-started) instructions to
get the Swift sources and then execute the `build-script` using swiftpm preset:

### macOS

```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA
```

### Linux

```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
```

Once the build is complete, you should be able to run the swiftpm binaries from the build folder.

### Example

```bash
$> cd /tmp && mkdir hello && cd hello
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
```

8. Test the Swift Package Manager.

```bash
$> Utilities/bootstrap test
```

> ## Getting Started

Reviewer: rename title to "Running the build"

1. Pull the SwiftPM repository:
```bash
$> git clone https://github.com/apple/swift-package-manager.git
```
2. Run your first build:
```bash
$> cd swift-package-manager
$> swift build
```
Make sure the build did not fail.  

A `.build/` folder will be generated and it should have inside a similar structure (including build binaries):
```bash
artifacts/
checkouts/
debug/
repositories
x86_64-apple-macosx
debug.yaml
manifest.db
workspace-state.json 
```
Binaries (in the example above) are in `x86_64-apple-macosx/`.  
If you need to build the generated binaries, run `swift-build` in inside `.build/`:
```bash
./.build/x86_64-apple-macosx/debug/swift-build
```

## Running Tests
### Run all tests.
```bash

> $> swift test --parallel

Reviewer: remove "--parallel"

```
### Run a single test.

> ```bash

Reviewer: Add short explanation above:

`For example, to only run the PackageGraphTests.DependencyResolverTests/testBasics`

$> swift test --filter PackageGraphTests.DependencyResolverTests/testBasics
```

> ### Run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases.

Reviewer: this should not be a title, its an example. replace with

`For example, to run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases.`

```bash
$> swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
```
Note: PackageDescription v4 is not available when developing using this method.

This method can also used be used for performance testing. Use the following command run SwiftPM's performance tests:

```bash
$> export TSC_ENABLE_PERF_TESTS=1
$> swift test -c release --filter PerformanceTests
```

> For example, testing basic functionality using the newly created binaries 

Reviewer: rephrase this sentence: `To test SwiftPM functionality using the newly created binaries, you can invoke them from the .build directory. For example:`

```bash
$> cd /tmp && mkdir hello && cd hello
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
```

## Testing on Linux with Docker
For contributors on macOS who need to test on Linux, install Docker and use the
following commands:

```bash
$> Utilities/Docker/docker-utils build # will build an image with the latest Swift snapshot
$> Utilities/Docker/docker-utils bootstrap # will bootstrap SwiftPM on the Linux container
$> Utilities/Docker/docker-utils run bash # to run an interactive Bash shell in the container
$> Utilities/Docker/docker-utils swift-build # to run swift-build in the container
$> Utilities/Docker/docker-utils swift-test # to run swift-test in the container
$> Utilities/Docker/docker-utils swift-run # to run swift-run in the container
```

> ## Creating Pull Requests

Reviewer: * please add a .GITHUB PR template to codify these expectations, see https://github.com/apple/swift-metrics/tree/main/.github

* please add a .GITHUB issue template codify these expectations, see https://github.com/apple/swift-metrics/tree/main/.github

* Add note on code style and how to use swift-format before submitting PRs (we just merged https://github.com/apple/swift-package-manager/pull/3012)

* Add comment on patch size, number of commits per PR, etc

1. Fork: https://github.com/apple/swift-package-manager
2. Clone a working copy of your fork
3. Create a new branch
4. Make your code changes
5. Commit (include the Radar link or JIRA issue id in the commit message if possible and a description your changes)
6. Push the commit / branch to your fork
7. Make a PR from your fork / branch to `apple: main`

> 8. Leave a new comment to trigger smoke tests: `@swift-ci please smoke test`

Reviewer: remove number 8, only the team can do that not the PR author

# Contributing to Swift Package Manager
There are several types of contributions one can make. Bug fixes, documentation and enhancements that do not materially change the user facing semantics of Swift Package Manager should be submitted directly as PR.

Larger changes that do materially change the semantics of Swift Package Manager (e.g. changes to the manifest format or behavior) are required to go through [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md).

> To see how previous evolution decisions for SwiftPM have been made and have some direction for the development of future features please check out the [Community Proposal](Documentation/Internals/PackageManagerCommunityProposal.md).

Reviewer: Community Proposal -> Community Proposals

and the link should point to https://forums.swift.org/tag/packagemanager

For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/).

## Reporting issues

> Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  

Reviewer: Lets clean these links up:

* [SwiftPM JIRA Bug Tracker](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22)
* [Guide for filing quality bug reports](https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug)

JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).

## Development environment

> SwiftPM is typically built with a pre-existing versions of SwiftPM present on the system. Which can be obtained by installing the latest release of Xcode or a toolchain from swift.org. But there are have multiple ways to setup your development environment:

Reviewer: versions -> version

Reviewer: system. Which -> system, which

### Using Xcode (Easiest)

1. Install latest Xcode from https://developer.apple.com/xcode

> 2. Open SwiftPM's `Package.swift` manifest with the latest release (including betas) of Xcode.

Reviewer: 1. Install Xcode from https://developer.apple.com/xcode (including betas!).
2. Open SwiftPM's `Package.swift` manifest with Xcode.

### Using a Swift toolchain

1. Download a toolchain from https://swift.org/download/

> 

Reviewer: remove empty line

2. Confirm the toolchain is installed correctly:

*macOS*
```bash
$> export TOOLCHAINS=swift
```

Verify that we're able to find the swift compiler from the installed toolchain.
```bash
$> xcrun --find swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

*Linux*
```bash
$> export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
```
Verify that we're able to find the swift compiler from the installed toolchain.
```bash
$> which swift
/path/to/swift-toolchain/usr/bin/swift
$> swift package --version
Swift Package Manager - Swift 5.3.0
$> swift --version
Apple Swift version 5.3
```

3. Alternatively use tools like swiftenv that help manage toolchains versions.

### Using the Swift Compiler Build Script

Follow these instructions to get the Swift sources and then execute the build-script using swiftpm preset:

*macOS*
```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA
```

*Linux*
```bash
$> ./swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
```

> 3. Clone [llbuild](https://github.com/apple/swift-llbuild) beside the package manager directory.

Reviewer: fix the numbering to start with 1

    "swift-llbuild".

4. Clone [Yams](https://github.com/jpsim/yams) beside the package manager directory.

```bash
$> git clone https://github.com/jpsim/yams
```

5. Clone [swift-driver](https://github.com/apple/swift-driver) beside the package manager directory.

```bash
$> git clone https://github.com/apple/swift-driver
```

6. Clone [swift-argument-parser](https://github.com/apple/swift-argument-parser) beside the package manager directory and check out tag with the [latest version](https://github.com/apple/swift-argument-parser/tags).

For example, if the latest tag is 0.3.1:
```bash
$> git clone https://github.com/apple/swift-argument-parser --branch 0.3.1
```

7. Build the Swift Package Manager.

```bash
$> cd swiftpm
$> Utilities/bootstrap build
```

 Note: The bootstrap script requires having [CMake](https://cmake.org/) and [Ninja](https://ninja-build.org/) installed. Please refer to the [Swift project repo](https://github.com/apple/swift/blob/master/README.md#macos) for installation instructions.

This command builds the Package Manager inside the `.build/` directory.
    Run the bootstrap script to rebuild after making a change to the source
    code.

> <a id="self-hosting">*_Self Hosting_*</a>:  

Reviewer: L115-175 are redundant, remove them

$> Utilities/Docker/docker-utils run bash # to run an interactive Bash shell in the container
$> Utilities/Docker/docker-utils swift-build # to run swift-build in the container
$> Utilities/Docker/docker-utils swift-test # to run swift-test in the container
$> Utilities/Docker/docker-utils swift-run # to run swift-run in the container
```

## Creating Pull Requests
1. Fork: https://github.com/apple/swift-package-manager
2. Clone a working copy of your fork
3. Create a new branch
4. Make your code changes
5. Try to keep your changes (when possible) below 200 lines of code.
6. We use [SwiftFormat](https://www.github.com/nicklockwood/SwiftFormat) to enforce code style. Please install and run SwiftFormat before submitting your PR.
7. Commit (include the Radar link or JIRA issue id in the commit message if possible and a description your changes). Try to have only 1 commit in your PR (but, of course, if you add changes that can be helpful to be kept aside from the previous commit, make a new commit for them).
8. Push the commit / branch to your fork
9. Make a PR from your fork / branch to `apple: main`

> 10. While creating your PR, make sure to add one of the [PR Templates](https://github.com/apple/swift-package-manager/tree/main/.github) in your PR message depending on the reasoning of your PR.

Reviewer: While creating your PR, make sure to follow the PR Template providing information about the motivation and highlighting the changes.

11. Reviewers are going to be automatically added to your PR
12. Merge pull request when you received approval from the reviewers (one or more)

## Using Continuous Integration
SwiftPM uses [swift-ci](https://ci.swift.org) infrastructure for its continuous integration testing. The bots can be triggered on pull-requests if you have commit access. Otherwise, ask

> one of the code owners to trigger them for you. The following commands are supported:

Reviewer: line break here seems odd?

```
@swift-ci please smoke test
```

Run tests with the trunk compiler and other projects. This is **required** before
a pull-request can be merged.

```

> @swift-ci please smoke test self hosted

Reviewer: @swift-ci please smoke test

environment variables to control which compiler to use for these operations.

`SWIFT_EXEC_MANIFEST`: This variable controls which compiler to use for parsing
`Package.swift` manifest files. The lookup order for the manifest compiler is:
`SWIFT_EXEC_MANIFEST`, `swiftc` adjacent to the `swiftpm` binaries, then `SWIFT_EXEC`

`SWIFT_EXEC`: This variable controls which compiler to use for compiling Swift
sources. The lookup order for the sources' compiler is: `SWIFT_EXEC`, then `swiftc` adjacent
to `swiftpm` binaries. This is also useful for Swift compiler developers when they
want to use a debug compiler with SwiftPM.

```bash
$> SWIFT_EXEC=/path/to/my/built/swiftc swift build
```

## Overriding the Path to the Runtime Libraries
SwiftPM computes the path of its runtime libraries relative to where it is
installed. This path can be overridden by setting the environment variable
`SWIFTPM_PD_LIBS` to a directory containing the libraries, or a colon-separated list of
absolute search paths. SwiftPM will choose the first
path which exists on disk. If none of the paths are present on disk, it will fall
back to built-in computation.

## Making changes in TSC targets

> All targets with the prefix TSC define the interface for the tools support core. Those APIs might be used in other projects as well and need to be updated in this repository by copying their sources directories to the TSC repository. The repository can be found [here](https://github.com/apple/swift-tools-support-core).

Reviewer: SwiftPM uses [Tools Support Core](https://github.com/apple/swift-tools-support-core) (aka TSC) for many of its general purpose utilities. Changes in SwiftPM often require changes in TSC first. To coordinate changes, open a PR against TSC first, then a second one against SwiftPM pulling the correct TSC version.

## Community and Support
If you want to connect with the Swift community you can:
* Use Swift Forums: [https://forums.swift.org/c/development/SwiftPM](https://forums.swift.org/c/development/SwiftPM)
* Contact the CODEOWNERS: https://github.com/apple/swift-package-manager/blob/main/CODEOWNERS

> (mailing lists are usually the best place to go for help: [code-owners@swift.org](mailto:code-owners@swift.org), [conduct@swift.org](mailto:conduct@swift.org), [swift-infrastructure@swift.org](mailto:swift-infrastructure@swift.org)

Reviewer: remove this line

To see how previous evolution decisions for SwiftPM have been made and have some direction for the development of future features please check out the [Community Proposal](Documentation/Internals/PackageManagerCommunityProposal.md).  

For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/)  

## Reporting issues
Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  
JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).

## Development environment
If you are contributing using macOS, the best option is to use Xcode to build and run test SwiftPM.  
You have multiple ways to setup your development environment, here we will focus on two:  
* [Using Xcode](#using-xcode)
* [Using a standalone Swift toolchain](#using-standalone)
* [Using a Trunk Snapshot](#using-trunk-snapshot)
* [Self Hosting](#self-hosting)

> * [Using the Swift Compiler Build Script](#swift-compiler-build-script)

Reviewer: This section needs further work. There is no material difference between "Using a standalone Swift toolchain", "Using a Trunk Snapshot" and "Self Hosting". I suggest the following text:

```markdown
##Development environment

SwiftPM is typically built with a pre-existing versions of SwiftPM present on the system. Which can be obtained by installing the latest release of Xcode or a toolchain from swift.org. But there are have multiple ways to setup your development environment:

### Using Xcode (Easiest)

1. Install latest Xcode from https://developer.apple.com/xcode
2. Open SwiftPM's `Package.swift` manifest with the latest release (including betas) of Xcode.

### Using a Swift toolchain

1. Download a toolchain from https://swift.org/download/

2. Confirm the toolchain is installed correctly:

macOS
$ export TOOLCHAINS=swift
# Verify that we're able to find the swift compiler from the installed toolchain.
$ xcrun --find swift
/Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift

Linux
$ export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
# Verify that we're able to find the swift compiler from the installed toolchain.
$ which swift
/path/to/swift-toolchain/usr/bin/swift

3. Alternatively use tools like swiftenv that help manage toolchains versions.

### Using the Swift Compiler Build Script

Follow these instructions to get the Swift sources and then execute the build-script using swiftpm preset:

macOS
$ ./swift/utils/build-script --preset=buildbot_swiftpm_macos_platform,tools=RA,stdlib=RA

Linux
$ ./swift/utils/build-script --preset=buildbot_swiftpm_linux_platform,tools=RA,stdlib=RA
Once the build is complete, you should be able to run the swiftpm binaries from the build folder.

Example

$ cd /tmp && mkdir hello && cd hello
$ /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$ /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
$ Utilities/bootstrap test
```

$ Utilities/bootstrap test
```

## Getting Started
1. Pull the SwiftPM repository:
```
git clone https://github.com/apple/swift-package-manager.git
```
2. Run your first build:
```
$> cd swift-package-manager
$> swift build
```
Make sure the build did not fail.  

A `.build/` folder will be generated and it should have inside a similar structure (including build binaries):
```
artifacts/
checkouts/
debug/
repositories
x86_64-apple-macosx
debug.yaml
manifest.db
workspace-state.json 
```
Binaries (in the example above) are in `x86_64-apple-macosx/`.  
If you need to build the generated binaries, run `swift-build` in inside `.build/`:
```
./.build/x86_64-apple-macosx/debug/swift-build

> ```

Reviewer: Add the following:

```
Testing basic functionality using the newly created binaries 

$ cd /tmp && mkdir hello && cd hello
$ /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$ /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
```

> 

Reviewer: Lets add a section here about testing:

```markdown
## Running Tests

# Run all tests.
$ swift test --parallel

# Run a single test.
$ swift test --filter PackageGraphTests.DependencyResolverTests/testBasics

# Run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases.
$ swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
Note: PackageDescription v4 is not available when developing using this method.

This method can also used be used for performance testing. Use the following command run SwiftPM's performance tests:

$ export TSC_ENABLE_PERF_TESTS=1
$ swift test -c release --filter PerformanceTests
```

> ## Using Continuous Integration

Reviewer: move this section after "Creating Pull Requests"

SwiftPM uses [swift-ci](https://ci.swift.org) infrastructure for its continuous integration testing. The bots can be triggered on pull-requests if you have commit access. Otherwise, ask
one of the code owners to trigger them for you. The following commands are supported:

```
@swift-ci please smoke test
```

Run tests with the trunk compiler and other projects. This is **required** before
a pull-request can be merged.

```
@swift-ci please smoke test self hosted
```

Run just the self-hosted tests. This has fast turnaround times so it can be used

> to get quick feedback.  

Reviewer: remove extra space

Note: Smoke tests are still required for merging pull-requests.

> ## Testing on Linux with Docker

Reviewer: move this section after the new "Running Tests"

back to built-in computation.

## Making changes in TSC targets
All targets with the prefix TSC define the interface for the tools support core. Those APIs might be used in other projects as well and need to be updated in this repository by copying their sources directories to the TSC repository. The repository can be found [here](https://github.com/apple/swift-tools-support-core).

> ## Commits / PRs

Reviewer: rename section "Creating Pull Requests" and move after "Testing on Linux with Docker"
